### ISSUE TITLE
The Material color API is not well suited for null safety

### ISSUE THREAD

**Issue by mit-mit** [flutter/flutter#71591]
The material color palette has a set of named colors with pre-set saturations of 50,100,200,300,400,500,600,700,800, and 900: https://material.io/resources/color/#!/?view.left=0&view.right=0

In the current Flutter API, this is encoded as a static named getters (e.g. `green`) on `Color` with an `[]` operator for getting the Color representing a given saturation (expressed as an `int`), e.g.: `Colors.green[400]` returns a `Color` that is Green with a saturation of 400.

After migrating Flutter to null safety, this causes issues when using these to set a non-null color, e.g.:

```dart
class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: const BoxDecoration(
        border: Border(
          top: BorderSide(width: 1.0, color: Colors.green[400]),
        ),
      ),
      child: const Text('Hello'),
    );
  }
}
```

=>

```
The argument type 'Color?' can't be assigned to the parameter type 'Color'.
```

It would be nice if this API could be redesigned so it returns something of type `Color`, not `Color?` for the known saturations.

**Comment by mit-mit** [comment]
cc @HansMuller

**Comment by jonahwilliams** [comment]
It also can't be used in const expressions either. We should replace the usage of the int -> Color map with dedicated final fields.

**Comment by mit-mit** [comment]
Ah, yes, we could simply add something like this?

```dart
class Colors {
  static const MaterialColor green = MaterialColor(
    _greenPrimaryValue,
    <int, Color>{
       50: Color(0xFFE8F5E9),
      100: Color(0xFFC8E6C9),
...
    },
  );

  static const green50 = Color(0xFFE8F5E9);
  static const green100 = Color(0xFFC8E6C9);
...
}
```

That would be non-breaking, and then we can deprecate the Color map later.

**Comment by filiph** [comment]
FWIW, I am in favor of the `Colors.green[500]` --> `Colors.green50` change.

It always seemed to me a bit dangerous to have a sparse map like that. The API makes it look like we can use it like `Colors.green[i]`, where `i` is interpolating between two shades. Which is not the case, and will crash. Which I guess the null safety feature makes very explicit now.

I like the _aesthetics_ of `Colors.green[500]`, but I don't like the usability. The new alternative _looks_ uglier (to me) but it's fool-proof.

**Comment by mit-mit** [comment]
cc @Hixie

**Comment by Hixie** [comment]
We already have this, `Colors.green.shade500` for example.

**Comment by mit-mit** [comment]
But doesn't look like the existing `shade---` colors are const either: https://github.com/flutter/flutter/issues/31351#issuecomment-618146805 :-(

**Comment by Hixie** [comment]
Yeah, they're backed by the map so they can't be const, unfortunately.

**Comment by rydmike** [comment]
I was reading this issue earlier today and came here to drop a reference to the discussion I had seen in this long outstanding issue/feature request https://github.com/flutter/flutter/issues/31351 from 2019 as well. Seems like @mit-mit beat me to it. 👍🏻  

Having colors with shades that are const would be a nice addition or option, as already discussed to some extent in https://github.com/flutter/flutter/issues/31351. If an implementation for this issue at the same time solves the old no const color shades issue then that would be welcome. 

On the other hand I do agree that all the different ways of referring to the colors can be confusing for newcomers, this was also mentioned in the discussion in https://github.com/flutter/flutter/issues/31351 as an objection to adding one more way. 

As was shown in https://github.com/flutter/flutter/issues/31351 the `Colors.green.shade500` style is clearly less frequently used in the framework (or it least it was back then) for referring to the shades. Seems to be rare to the extent that one might sometimes even forget the format exist.

If adding const colors and shades as proposed in https://github.com/flutter/flutter/issues/31351 would result in too many ways of referring to color shades, then deprecating the current .shadeNNN option over the [NNN] option would be my preference.

The [NNN] variant is nice to have sometimes, you can e.g. use a `for` iteration to create a complete Material like swatch from a single custom color, by doing something like this:

```dart
  MaterialColor createPrimarySwatch(Color color) {
    final List<double> strengths = <double>[0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9];
    final Map<int, Color> swatch = <int, Color>{};
    final int r = color.red, g = color.green, b = color.blue;
    for (final double strength in strengths) {
      final double ds = 0.5 - strength;
      swatch[(strength * 1000).round()] = Color.fromRGBO(
        r + ((ds < 0 ? r : (255 - r)) * ds).round(),
        g + ((ds < 0 ? g : (255 - g)) * ds).round(),
        b + ((ds < 0 ? b : (255 - b)) * ds).round(),
        1,
      );
    }
    return MaterialColor(color.value, swatch);
  }
```

Which gets you a complete Material primary color like swatch, where the passed in color becomes the color value and the "mid" [500] shade and the rest of the swatch [NNN] shade values are computed to create a primary like swatch. 

I used this approach before when the swatch was more of a thing in themes in order to create a complete computed swatch from a single given custom color. With the move to ColorScheme based themes, the swatch probably gets less usage in the framework and probably less in newer apps too. The above is still however a handy and useful way to create additional Material like shades for custom primary, primaryVariant, secondary and secondaryVariant colors in ColorScheme based themes too, if you need extra Material like shades of them, which you do in apps sometimes, while still keeping the colors shades related to the ColorScheme color.

**Comment by esmaeil-ahmadipour** [comment]
In null safty have this problem.

This problem was solved this way:

color: Colors.blue[100]!
